untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+
+<p>
+<details>
+<summary>Haskell and call by need</summary>
+
+In theory, Haskell strives to implement call-by-need.
+In practice, it is a little more complicated.
+The reason is simple -- the optimizer can make non-obvious transformations, which could either remove some duplicated computations or vice versa.
+
+***
+
+</details>
+</p>
+
+Expressions that are not reducible using the chosen evaluation strategy are said to be in the _normal form_.
+
+All of the strategies described here, except full beta-reduction, are normalizing, i.e. if a term has a normal form, it will eventually reduce to it.
+Furthermore, each term that has a normal form, has the same normal form under both strong evaluation strategies.
+Likewise for the weak evaluation strategies.
+
+Speaking of reductions, we'll also mention _eta_-reduction ($\eta$-reduction).
+The rule of eta-reduction says that $\lambda x. f\;x = f,$ if $x$ does not appear free in $f$.
+It boils down to the statement that two functions are the same if and only if they give the same result for all possible arguments.
+
+## Programming in pure untyped lambda calculus
+
+> When a pattern shows itself in tiles or on paper or in your mind and says, 'This is the mode of my repetition; in this manner I extend myself to infinity', it has already done so, it has already been infinite from the very first moment of its being; the potentiality and the actuality are one thing.
+> If two and two can be four then they actually are four, you can only perceive it, you have no part in making it happen by writing it down in numbers or telling it out in pebbles.
+>
+> -- Russell Hoban, Pilgermann, Pan, London, 1984.
+
+Now that we defined both syntax and semantics, we can talk about expressing various computations in the language of lambda calculus.
+It is not that different from using any other programming language.
+
+We run into a small issue, however.
+We don't have much to work with yet.
+We don't have booleans, numbers, tuples, or multivariate functions.
+No worries, though!
+We can represent all of that using only abstractions and applications!
+
+### Multivariate functions
+
+Any multivariate function is the same as multiple nested functions returning other functions.
+If you're familiar with functional programming, you might know such functions are called curried functions, named after Haskell Curry.
+In pure lambda calculus, then, all functions are curried.
+
+To simplify the notation, we will now introduce a shorthand syntax.
+A multivariate abstraction of $n$ parameters, declared as $\lambda x_1\;x_2\;\ldots\;x_n. t,$ is the same as $n$ nested abstractions $\lambda x_1. \lambda x_2. \ldots \lambda x_n. t.$
+
+Multivariate abstractions, naturally, can be partially applied, e.g. $(\lambda x\; y\; z. (x\;y)\;(x\;z))\;a\;b$ reduces to $\lambda z. (a\;b)\;(a\;z).$
+
+### Church booleans
+
+Let us now introduce booleans.
+We will follow Church here.
```

Comment:
Do you think it is clear what this sentence means? Like, I understood, but some might be confused that you mean that you will cover the Church encoding of booleans

Reply:
Or https://wiki.c2.com/?ChurchOfTheLeastFixedPoint.

Reply:
Changed to

> We will do it in the same way Church did in his original paper.

Although I was briefly tempted to add "the" before "Church" and call it a day :joy: